### ISSUE TITLE
Project graph construction takes a long time

### ISSUE THREAD

**Issue by prihoda** [dataversioncontrol/dvc#711]
Hi guys, creating the project graph takes minutes on my project with 54 stages. 

Vast majority of the time is spent in the output.py `Dependency.stage` function that is repeatedly called for each dependency:
```
        for stage in self.project.stages():
            for out in stage.outs:
                if self.path == out.path:
                    return stage
```

Any ideas how to improve that? I haven't had the chance to understand all the code yet.

**Comment by efiop** [comment]
Hi @prihoda ! Great point! Current implementation is not very efficient and we are aware of that and plan to look into performance improvements in the future. In the current issue can be easily solved by not re-collecting stages each time. I'll look into it today and will try to release the fix with 0.9.7. Thank you!

**Comment by prihoda** [comment]
Awesome, thanks! This makes my workflow so much faster
